src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs

```diff
@@ -85,28 +86,45 @@ public Block[] Process(Hash256 newBranchStateRoot, List<Block> suggestedBlocks,
     {
         if (suggestedBlocks.Count == 0) return [];
 
-        TxHashCalculator.CalculateInBackground(suggestedBlocks);
-        BlocksProcessing?.Invoke(this, new BlocksProcessingEventArgs(suggestedBlocks));
-
         /* We need to save the snapshot state root before reorganization in case the new branch has invalid blocks.
            In case of invalid blocks on the new branch we will discard the entire branch and come back to
            the previous head state.*/
         Hash256 previousBranchStateRoot = CreateCheckpoint();
         InitBranch(newBranchStateRoot);
+
+        Block suggestedBlock = suggestedBlocks[0];
+        CancellationTokenSource? cancellationTokenSource = null;
+        Task? preWarmTask = null;
+        // Start prewarming as early as possible
+        WaitForCacheClear();
+        bool skipPrewarming = preWarmer is null || suggestedBlock.Transactions.Length < 3;
+        if (!skipPrewarming)
+        {
+            (cancellationTokenSource, preWarmTask) = PreWarmTransactions(suggestedBlock, newBranchStateRoot);
+        }
+
+        BlocksProcessing?.Invoke(this, new BlocksProcessingEventArgs(suggestedBlocks));
+
         Hash256 preBlockStateRoot = newBranchStateRoot;
 
         bool notReadOnly = !options.ContainsFlag(ProcessingOptions.ReadOnlyChain);
         int blocksCount = suggestedBlocks.Count;
         Block[] processedBlocks = new Block[blocksCount];
-
-        Task? preWarmTask = null;
         try
         {
             for (int i = 0; i < blocksCount; i++)
             {
-                preWarmTask = null;
                 WaitForCacheClear();
-                Block suggestedBlock = suggestedBlocks[i];
+                suggestedBlock = suggestedBlocks[i];
+
+                skipPrewarming = preWarmer is null || suggestedBlock.Transactions.Length < 3;
+                // If cancelationTokenSource is not null it means we are in first iteration of loop
+                // and started prewarming at method entry, so don't start it again
+                if (cancellationTokenSource is null && !skipPrewarming)
+                {
+                    (cancellationTokenSource, preWarmTask) = PreWarmTransactions(suggestedBlock, preBlockStateRoot);
+                }
```

Comment:
This code is very similar to the one above, can we have a helper method?

Reply:
Moved into PreWarm method